tools/ansible/tasks/wsu/main.yaml

```diff
@@ -2,29 +2,131 @@
+
+    # Changelog url to lookup SHA for kube node binaries
+    # Example: https://github.com/kubernetes/kubernetes/blob/release-1.17/CHANGELOG/CHANGELOG-1.17.md
+    - name: Get kubernetes changelog url
+      shell: |
+        echo "https://github.com/kubernetes/kubernetes/blob/release-"\
+        "{{ kubernetes_version_without_patch.stdout }}/CHANGELOG/CHANGELOG-"\
+        "{{ kubernetes_version_without_patch.stdout }}.md"
+      register: changelog
+
+    - name: Get kube node binary SHA
+      shell: |
+        curl -s "{{ changelog.stdout }}" | \
+        grep -A 1 "{{ kubernetes_version.stdout }}" | \
+        grep -A 1 node-windows | tail -n 1 | sed -e 's/.*<code>\(.*\)<\/code>.*/\1/'
+      register: kube_node_sha
+
+    # OpenShift cluster version
+    - name: Register cluster version
+      shell: "echo {{ map_kubernetes_to_openshift_version[kubernetes_version_without_patch.stdout] }}"
+      register: cluster_version
 
-        - name: Move WMCB to temporary directory
-          command: mv "{{ project_root }}/{{ wmcb_exe }}" "{{ tmp_dir.path }}/{{ wmcb_exe }}"
+    # curl -s https://api.github.com/repos/openshift/windows-machine-config-operator/releases gets the list of all
+    # releases, we use the cluster version from previous ansible task to select the release we need.
+    - name: Get release
+      shell: |
+        RELEASES="$(curl -s https://api.github.com/repos/openshift/windows-machine-config-operator/releases)"
+        TAGNAME="$(echo $RELEASES | jq '.[]' | jq '.tag_name' | \
+        grep -i {{ cluster_version.stdout }} |\
+        sort -r | awk 'NR==1')"
+        echo $RELEASES | \
+        jq '.[] | select (.tag_name == '$TAGNAME')'
+      register: release
+
+    - name: Gather required files
+      block:
+        - name: Build WMCB
+          when: build_wmcb is defined and build_wmcb
+          block:
+            - name: Build WMCB
+              make:
+                target: build
+                chdir: "{{ project_root }}"
+
+            - name: Move WMCB to temporary directory
+              command: mv "{{ project_root }}/{{ wmcb_exe }}" "{{ tmp_dir.path }}/{{ wmcb_exe }}"
+
+        - name: Download WMCB binary
+          when: not build_wmcb or build_wmcb is not defined
+          block:
+            # The wmcb release tags are supposed to conform to the following semver format:
+            # v<openshift-major-version>.<openshift-minor-version>.<wmcb_build_number>-<release_type>. For example, v4.4-1-alpha.
+            # For a <openshift-major-version>.<openshift-minor-version>, there can be many wmcb builds available, for
+            # example, OpenShift 4.4 can map to v4.4.1-alpha and v4.4.2-alpha, we will always pick up the latest build available
+            # which is v4.4.2-alpha,
+            # In future, when we cut beta and GA releases, the playbook should always go for GA version when available.
+            - name: Get wmcb url
+              shell: |
```

Comment:
Have you tried [`get_url` ](https://docs.ansible.com/ansible/latest/modules/get_url_module.html)module in ansible, it seems you can use checksum as parameter to it, which might save steps for us, like checking for check sum, if ansible is already doing that for us.

Reply:
That is correct - `get_url` along with `win_copy` modules should ensure checksums are checked. I mimicked the SHA check of WMCB from that of hybrid-overlay, thinking it should cover the edge case of redirects - if that situation ever arises.

Reply:
I don't follow you reasoning. Are you saying that `get_url` and `win_copy` do support redirects?

Reply:
I am not 100% sure if `get_url` doesn't automatically follow redirects (although `win_get_url` doesn't - that's why we have added the option [here](https://github.com/openshift/windows-machine-config-bootstrapper/blob/master/tools/ansible/tasks/wsu/main.yaml#L135). While exploring with [checksum](https://docs.ansible.com/ansible/latest/modules/get_url_module.html#parameter-checksum) parameter of `get_url`, I discovered that it doesn't work the way we want it to. It uses this checksum so it can skip download if the file is already present, but if the file downloaded doesn't meet the checksum provided, then it fails with "The checksum for <file_downloaded> did not match <expected_checksum>; it was <actual_checksum>". If we are okay with this, I can make the changes.

Reply:
Isn't this the behavior that we want?

Reply:
> It uses this checksum so it can skip download if the file is already present,

What's wrong with `force` option if you want the download to happen always.

> but if the file downloaded doesn't meet the checksum provided, then it fails with "The checksum for <file_downloaded> did not match <expected_checksum>; it was <actual_checksum>". If we are okay with this, I can make the changes.

Don't we fail, if the checksum fails even today?

Reply:
After discussing it with @ravisantoshgudimetla , I have decided to use `get_url` module's checksum parameter, which eliminates the need to check the checksum on windows instance. Implemented for both kube node binaries and downloaded wmcb

Reply:
> which eliminates the need to check the checksum on windows instance

Why? What if there is an error when transferring the file to the Windows instance?

Reply:
`win_copy` will ensure they are copied over correctly or fail

Reply:
Please define "correctly" here. It is not clear how and when the cheksum is done as per the [docs](https://docs.ansible.com/ansible/latest/modules/win_copy_module.html).

Reply:
I added some comments with `Download wmcb binary` task and `Copy required files to Windows host` task that should ensure the integrity of downloaded as well as copied file.
It is implicitly defined in the [docs for force parameter](https://docs.ansible.com/ansible/latest/modules/win_copy_module.html#parameter-force) that default run is doing the checksum